internal/infrastructure/kubernetes/service.go

```diff
@@ -0,0 +1,118 @@
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/intstr"
+
+	"github.com/envoyproxy/gateway/internal/ir"
+)
+
+const (
+	// envoyServiceHTTPPort is the HTTP port number of the Envoy service.
+	envoyServiceHTTPPort = int32(80)
+	// envoyServiceHTTPSPort is the HTTPS port number of the Envoy service.
+	envoyServiceHTTPSPort = int32(443)
+)
+
+// createServiceIfNeeded creates a Service based on the provided infra, if
+// it doesn't exist in the kube api server.
+func (i *Infra) createServiceIfNeeded(ctx context.Context, infra *ir.Infra) error {
+	current, err := i.getService(ctx, infra)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			svc, err := i.createService(ctx, infra)
+			if err != nil {
+				return err
+			}
+			if err := i.addResource(svc); err != nil {
+				return err
+			}
+			return nil
+		}
+		return err
+	}
+
+	if err := i.addResource(current); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// getService gets the Service from the kube api for the provided infra.
+func (i *Infra) getService(ctx context.Context, infra *ir.Infra) (*corev1.Service, error) {
+	ns := i.Namespace
+	name := infra.Proxy.ObjectName()
+	key := types.NamespacedName{
+		Namespace: ns,
+		Name:      name,
+	}
+	svc := new(corev1.Service)
+	if err := i.Client.Get(ctx, key, svc); err != nil {
+		return nil, fmt.Errorf("failed to get service %s/%s: %w", ns, name, err)
+	}
+
+	return svc, nil
+}
+
+// expectedService returns the expected Service based on the provided infra.
+func (i *Infra) expectedService(infra *ir.Infra) *corev1.Service {
+	var ports []corev1.ServicePort
+	for _, listener := range infra.Proxy.Listeners {
+		for _, port := range listener.Ports {
+			// Set the target port based on the protocol of the IR port.
+			target := intstr.IntOrString{IntVal: envoyHTTPPort}
+			if port.Protocol == ir.HTTPSProtocolType {
+				target = intstr.IntOrString{IntVal: envoyHTTPSPort}
+			}
+			p := corev1.ServicePort{
+				Name:       port.Name,
+				Protocol:   "TCP",
```

Comment:
Prefer to use constants like [`corev1.ProtocolTCP`](https://pkg.go.dev/k8s.io/api/core/v1#Protocol) instead of magic strings.